File: client/src/functions/crypto.ts
import sodium from "libsodium-wrappers";
import { Buffer } from "buffer";

const minAdditionalChars = 20;
const maxAdditionalChars = 100;

async function randomIntInRange(): Promise<number> {
    await sodium.ready; // Initialize sodium
    const range = maxAdditionalChars - minAdditionalChars + 1;
    const randomBytes = sodium.randombytes_buf(4);
    const randNum = new DataView(randomBytes.buffer).getUint32(0, true);
    return minAdditionalChars + (randNum % range);
}

interface PaddedString {
    content: string;
    padding: string;
}

async function randomLenPadString(input: string): Promise<string> {
    const paddingLen = await randomIntInRange();
    const padding = "A".repeat(paddingLen);
    const padded: PaddedString = {
        content: input,
        padding,
    };
    return JSON.stringify(padded);
}

function removePaddingFromPaddedString(input: string): string {
    const padded: PaddedString = JSON.parse(input);
    return padded.content;
}

function base64decodeStringToBytes(base64: string) {
    const binString = atob(base64);
    return Uint8Array.from(binString, (m) => m.charCodeAt(0));
}

function base64EncodeBytesToString(bytes: Uint8Array) {
    const binString = Array.from(bytes, (byte) => String.fromCharCode(byte)).join("");
    return btoa(binString);
}

export async function generateKey(): Promise<string> {
    await sodium.ready; // Initialize sodium

    return base64EncodeBytesToString(await sodium.randombytes_buf(sodium.crypto_secretbox_KEYBYTES));
}

export async function encryptWithPSK(plaintext: string, psk: string): Promise<string> {
    const pskbuf = Buffer.from(base64decodeStringToBytes(psk));

    await sodium.ready; // Initialize sodium

    // Generate a nonce (random number)
    const nonce = sodium.randombytes_buf(sodium.crypto_secretbox_NONCEBYTES);

    // Encrypt the plaintext with the pre-shared key (psk)
    const ciphertext = sodium.crypto_secretbox_easy(Buffer.from(await randomLenPadString(plaintext), "utf8"), nonce, pskbuf);

    // Return the nonce and ciphertext as a base64 encoded string
    return `${base64EncodeBytesToString(nonce)}:${base64EncodeBytesToString(ciphertext)}`;
}

export async function decryptWithPSK(encryptedData: string, psk: string): Promise<string> {
    await sodium.ready; // Initialize sodium

    const pskbuf = Buffer.from(base64decodeStringToBytes(psk));

    const parts = encryptedData.split(":");
    if (parts.length !== 2) {
        throw new Error("Invalid encrypted data format.");
    }

    // Decode base64 encoded nonce and ciphertext
    const nonce = Buffer.from(base64decodeStringToBytes(parts[0]));
    const ciphertext = Buffer.from(base64decodeStringToBytes(parts[1]));

    // Decrypt the data using the pre-shared key (psk)
    const decrypted = sodium.crypto_secretbox_open_easy(ciphertext, nonce, pskbuf);

    if (!decrypted) {
        throw new Error("Decryption failed.");
    }

    return removePaddingFromPaddedString(Buffer.from(decrypted).toString("utf8"));
}

File: server/src/crypto.rs
use base64::{
    alphabet,
    engine::{self, general_purpose, DecodePaddingMode},
    Engine as _,
};
use serde::{Deserialize, Serialize};
use sodiumoxide::crypto::secretbox;
use sodiumoxide::randombytes::randombytes_into;

const MIN_ADDITIONAL_CHARS: u32 = 20;
const MAX_ADDITIONAL_CHARS: u32 = 100;

pub fn init_rng() {
    sodiumoxide::init().unwrap();
}

fn random_int_in_range() -> u32 {
    let range = MAX_ADDITIONAL_CHARS - MIN_ADDITIONAL_CHARS + 1;
    let mut buf = [0u8; 4];
    randombytes_into(&mut buf);
    let rand_num = u32::from_le_bytes(buf);
    MIN_ADDITIONAL_CHARS + (rand_num % range)
}

#[derive(Serialize, Deserialize)]
struct PaddedString {
    content: String,
    padding: String,
}

fn random_len_pad_string(input: &str) -> String {
    let padding_len = random_int_in_range();
    let padding = "A".repeat(padding_len as usize);
    let padded = PaddedString {
        content: input.to_string(),
        padding,
    };
    serde_json::to_string(&padded).unwrap()
}

fn remove_padding_from_padded_string(input: &str) -> String {
    let padded: PaddedString = serde_json::from_str(input).unwrap();
    padded.content
}

const BASE_64_ENGINE: engine::GeneralPurpose = engine::GeneralPurpose::new(
    &alphabet::STANDARD,
    general_purpose::PAD
        .with_encode_padding(true)
        .with_decode_padding_mode(DecodePaddingMode::Indifferent),
);

pub fn base_64_encode_bytes_to_string(input: &Vec<u8>) -> String {
    BASE_64_ENGINE.encode(input)
}

pub fn base_64_decode_string_to_bytes(input: &str) -> Vec<u8> {
    BASE_64_ENGINE.decode(input).unwrap()
}

pub fn generate_key() -> String {
    base_64_encode_bytes_to_string(&secretbox::gen_key().as_ref().to_vec())
}

pub fn encrypt_with_psk(plaintext: &str, psk: &str) -> String {
    // Generate a nonce (random number) for encryption
    let nonce = secretbox::gen_nonce();

    // Encrypt the plaintext with a pre-shared key (psk) using secretbox (Sodium's authenticated encryption)
    let key = secretbox::Key::from_slice(&*base_64_decode_string_to_bytes(psk))
        .ok_or("Invalid PSK")
        .unwrap();
    let message_bytes_utf8 = random_len_pad_string(plaintext).as_bytes().to_vec();
    let ciphertext = secretbox::seal(&message_bytes_utf8, &nonce, &key);

    // Return the nonce and ciphertext as a base64 encoded string
    format!(
        "{}:{}",
        base_64_encode_bytes_to_string(&nonce.as_ref().to_vec()),
        base_64_encode_bytes_to_string(&ciphertext)
    )
}

pub fn decrypt_with_psk(encrypted_data: &str, psk: &str) -> String {
    // Decode the base64-encoded nonce and ciphertext
    let parts: Vec<&str> = encrypted_data.split(':').collect();
    let nonce = base_64_decode_string_to_bytes(&parts[0]);
    let ciphertext = base_64_decode_string_to_bytes(&parts[1]);

    // Decrypt the data using the pre-shared key (psk) and nonce
    let key = secretbox::Key::from_slice(&*base_64_decode_string_to_bytes(psk))
        .ok_or("Invalid PSK")
        .unwrap();
    let decrypted = secretbox::open(
        &ciphertext,
        &secretbox::Nonce::from_slice(&nonce).unwrap(),
        &key,
    )
    .unwrap();

    remove_padding_from_padded_string(&String::from_utf8_lossy(&decrypted))
}
